File: scripts/resolve-karcher-urls.mjs
/**
 * resolve-karcher-urls.mjs
 * 
 * Reads KARCHER-Grid_view.csv, follows the search redirect for each product
 * to get the real product page URL, then stores it in Supabase as manufacturer_url.
 * 
 * The CSV has URLs like:
 *   https://www.kaercher.com/ro/cautare-rezultate.html?query=1.004-062.0
 * 
 * Following the redirect or scraping the search results page gives us:
 *   https://www.kaercher.com/ro/professional/.../ab-20-ec-10040620.html
 * 
 * Usage:
 *   node --env-file=.env.local scripts/resolve-karcher-urls.mjs KARCHER-Grid_view.csv
 *   node --env-file=.env.local scripts/resolve-karcher-urls.mjs KARCHER-Grid_view.csv --limit 20
 *   node --env-file=.env.local scripts/resolve-karcher-urls.mjs KARCHER-Grid_view.csv --resume
 */

import { createClient } from '@supabase/supabase-js'
import { createReadStream } from 'fs'
import { createInterface } from 'readline'
import { resolve } from 'path'

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY

if (!SUPABASE_URL || !SUPABASE_KEY) { console.error('Missing SUPABASE env vars'); process.exit(1) }

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY)
const csvPath = resolve(process.argv[2] || 'KARCHER-Grid_view.csv')
const args = process.argv.slice(3)
const limitArg = args.includes('--limit') ? parseInt(args[args.indexOf('--limit') + 1]) : null
const resumeMode = args.includes('--resume')

// ── CSV Parser ────────────────────────────────────────────────────────────────

function parseCSVLine(line) {
  const result = []
  let current = '', inQuotes = false
  for (let i = 0; i < line.length; i++) {
    const ch = line[i]
    if (ch === '"') {
      if (inQuotes && line[i+1] === '"') { current += '"'; i++ }
      else inQuotes = !inQuotes
    } else if (ch === ',' && !inQuotes) {
      result.push(current); current = ''
    } else current += ch
  }
  result.push(current)
  return result
}

async function parseCSV(filePath) {
  const rows = []
  const rl = createInterface({ input: createReadStream(filePath), crlfDelay: Infinity })
  let headers = null
  for await (const line of rl) {
    const cols = parseCSVLine(line)
    if (!headers) { headers = cols.map(h => h.replace(/^\uFEFF/, '').trim()) }
    else { const row = {}; headers.forEach((h, i) => { row[h] = (cols[i] || '').trim() }); rows.push(row) }
  }
  return rows
}

// ── URL Resolution ────────────────────────────────────────────────────────────

const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
  'Accept': 'text/html,application/xhtml+xml',
  'Accept-Language': 'ro-RO,ro;q=0.9',
}

async function resolveProductUrl(searchUrl, sku) {
  const numericSlug = sku.replace(/[.\-]/g, '')
  
  // Strategy 1: Follow the search URL — Kärcher search page has product links in HTML
  try {
    const controller = new AbortController()
    setTimeout(() => controller.abort(), 12000)
    const res = await fetch(searchUrl, { signal: controller.signal, headers: HEADERS })
    
    if (res.ok) {
      const html = await res.text()
      // Look for product URL containing the numeric slug
      const regex = new RegExp(`href="(/ro/[^"]+${numericSlug}[^"]*\\.html)"`, 'i')
      const match = html.match(regex)
      if (match) {
        return `https://www.kaercher.com${match[1]}`
      }
      
      // Also check if we were redirected to the product page directly
      if (res.url && res.url.includes(numericSlug) && res.url.includes('.html')) {
        return res.url
      }
      
      // Try to find any product link in search results
      const anyProduct = html.match(/href="(\/ro\/(?:professional|home-garden|accesorii)\/[^"]+\.html)"/)
      if (anyProduct && !anyProduct[1].includes('search')) {
        return `https://www.kaercher.com${anyProduct[1]}`
      }
    }
  } catch (e) {
    // timeout or network error
  }

  // Strategy 2: Try Kärcher's JSON search API
  try {
    const controller = new AbortController()
    setTimeout(() => controller.abort(), 10000)
    const apiUrl = `https://www.kaercher.com/api/startpage/search?q=${encodeURIComponent(sku)}&country=ro&language=ro&num=5`
    const res = await fetch(apiUrl, {
      signal: controller.signal,
      headers: { ...HEADERS, 'Accept': 'application/json', 'X-Requested-With': 'XMLHttpRequest', 'Referer': 'https://www.kaercher.com/ro/' }
    })
    if (res.ok) {
      const text = await res.text()
      if (text && text.trim().startsWith('{')) {
        const data = JSON.parse(text)
        const items = [...(data?.suggestions||[]), ...(data?.products||[]), ...(data?.results||[])]
        for (const item of items) {
          const url = item?.url || item?.link || ''
          if (url && url.includes(numericSlug)) {
            return url.startsWith('http') ? url : `https://www.kaercher.com${url}`
          }
        }
      }
    }
  } catch (e) {}

  // Strategy 3: Direct path construction
  const nameSlug = numericSlug // simplest fallback
  const paths = [
    `https://www.kaercher.com/ro/accesorii/${nameSlug}.html`,
    `https://www.kaercher.com/ro/search/?q=${numericSlug}`,
  ]
  
  return null // not found
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  console.log(`📂 Reading: ${csvPath}`)
  const rows = await parseCSV(csvPath)
  console.log(`📊 ${rows.length} rows in CSV`)

  // If resume mode, get already-resolved SKUs from Supabase
  let resolvedSkus = new Set()
  if (resumeMode) {
    const { data } = await supabase
      .from('products')
      .select('sku')
      .not('manufacturer_url', 'is', null)
      .eq('brand_name', 'Karcher')
    resolvedSkus = new Set((data || []).map(r => r.sku))
    console.log(`⏭  Skipping ${resolvedSkus.size} already resolved`)
  }

  // Filter rows
  let toProcess = rows.filter(r => {
    const url = r['Manufacturer URL'] || ''
    return r['SKU'] && url.includes('kaercher.com') && !resolvedSkus.has(r['SKU'])
  })

  if (limitArg) toProcess = toProcess.slice(0, limitArg)
  console.log(`\n🔍 Resolving ${toProcess.length} product URLs...\n`)

  const results = { found: 0, notFound: 0, error: 0 }

  for (let i = 0; i < toProcess.length; i++) {
    const row = toProcess[i]
    const sku = row['SKU']
    const searchUrl = row['Manufacturer URL']
    
    process.stdout.write(`[${i+1}/${toProcess.length}] ${sku}... `)

    const productUrl = await resolveProductUrl(searchUrl, sku)

    if (productUrl) {
      // Save to Supabase
      const { error } = await supabase
        .from('products')
        .update({ manufacturer_url: productUrl })
        .eq('sku', sku)
      
      if (error) {
        console.log(`✗ DB error: ${error.message}`)
        results.error++
      } else {
        console.log(`✓ ${productUrl.split('/ro/')[1]?.slice(0, 60)}`)
        results.found++
      }
    } else {
      console.log(`⚠ not found`)
      results.notFound++
    }

    // Polite delay
    if (i < toProcess.length - 1) await new Promise(r => setTimeout(r, 1500))
  }

  console.log(`\n── Summary ───────────────────────────`)
  console.log(`✓ Resolved: ${results.found}`)
  console.log(`⚠ Not found: ${results.notFound}`)
  console.log(`✗ Errors: ${results.error}`)
}

main().catch(console.error)

File: scripts/resolve-karcher-urls.py
#!/usr/bin/env python3
"""
resolve-karcher-urls.py
========================
Reads KARCHER-Grid_view.csv, finds the real product URL on kaercher.com/ro
for each SKU using Bing search, saves manufacturer_url to Supabase.

Usage:
  python3 scripts/resolve-karcher-urls.py KARCHER-Grid_view.csv
  python3 scripts/resolve-karcher-urls.py KARCHER-Grid_view.csv --limit 20
  python3 scripts/resolve-karcher-urls.py KARCHER-Grid_view.csv --resume
  python3 scripts/resolve-karcher-urls.py KARCHER-Grid_view.csv --dry-run
"""

import csv, sys, os, re, time, argparse, requests
from dotenv import load_dotenv

load_dotenv('.env.local')

SUPABASE_URL = os.environ['NEXT_PUBLIC_SUPABASE_URL']
SUPABASE_KEY = os.environ.get('SUPABASE_SERVICE_KEY') or os.environ['NEXT_PUBLIC_SUPABASE_ANON_KEY']

HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml',
    'Accept-Language': 'ro-RO,ro;q=0.9,en;q=0.8',
}

def sku_to_slug(sku):
    return re.sub(r'[.\-]', '', sku)

def find_product_url(sku):
    numeric = sku_to_slug(sku)
    
    # Strategy 1: Bing search — site:kaercher.com/ro {numeric}
    try:
        res = requests.get(
            f'https://www.bing.com/search?q=site:kaercher.com/ro+{numeric}&setlang=ro',
            headers=HEADERS, timeout=12
        )
        if res.ok:
            # Find kaercher.com/ro URL containing the numeric slug
            match = re.search(
                rf'https://www\.kaercher\.com/ro/[^"\'<>\s]+{numeric}[^"\'<>\s]*\.html',
                res.text
            )
            if match:
                return match.group(0).split('"')[0]  # clean any trailing chars
    except Exception as e:
        pass
    
    # Strategy 2: DuckDuckGo
    try:
        res = requests.get(
            f'https://html.duckduckgo.com/html/?q=site:kaercher.com/ro+{numeric}',
            headers=HEADERS, timeout=12
        )
        if res.ok:
            match = re.search(
                rf'https://www\.kaercher\.com/ro/[^"\'<>\s]+{numeric}[^"\'<>\s]*\.html',
                res.text
            )
            if match:
                return match.group(0).split('"')[0]
    except:
        pass

    # Strategy 3: Try kaercher.com/ro/search directly
    try:
        res = requests.get(
            f'https://www.kaercher.com/ro/search/?q={numeric}',
            headers=HEADERS, timeout=12
        )
        if res.ok:
            match = re.search(
                rf'href="(/ro/[^"]+{numeric}[^"]*\.html)"',
                res.text
            )
            if match:
                return f'https://www.kaercher.com{match.group(1)}'
    except:
        pass

    # Strategy 4: Try common flat paths
    for base in ['accesorii', 'professional', 'home-garden']:
        try:
            url = f'https://www.kaercher.com/ro/{base}/{numeric}.html'
            r = requests.head(url, headers=HEADERS, timeout=8, allow_redirects=True)
            if r.status_code == 200:
                return r.url
        except:
            pass
    
    return None

def supabase_get(path):
    r = requests.get(
        f'{SUPABASE_URL}/rest/v1/{path}',
        headers={
            'apikey': SUPABASE_KEY,
            'Authorization': f'Bearer {SUPABASE_KEY}',
        }
    )
    return r.json()

def supabase_update(sku, manufacturer_url):
    r = requests.patch(
        f'{SUPABASE_URL}/rest/v1/products?sku=eq.{requests.utils.quote(sku)}',
        headers={
            'apikey': SUPABASE_KEY,
            'Authorization': f'Bearer {SUPABASE_KEY}',
            'Content-Type': 'application/json',
            'Prefer': 'return=minimal',
        },
        json={'manufacturer_url': manufacturer_url}
    )
    return r.status_code in (200, 204)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('csv_file')
    parser.add_argument('--limit', type=int)
    parser.add_argument('--resume', action='store_true')
    parser.add_argument('--dry-run', action='store_true')
    args = parser.parse_args()

    # Read CSV
    with open(args.csv_file, encoding='utf-8-sig') as f:
        rows = list(csv.DictReader(f))
    print(f'📊 {len(rows)} rows in CSV')

    # Get already resolved SKUs if resume
    resolved = set()
    if args.resume:
        data = supabase_get('products?select=sku&brand_name=eq.Karcher&manufacturer_url=not.is.null')
        resolved = {r['sku'] for r in data}
        print(f'⏭  Skipping {len(resolved)} already resolved')

    to_process = [r for r in rows if r['SKU'] and r['SKU'] not in resolved]
    if args.limit:
        to_process = to_process[:args.limit]
    
    print(f'🔍 Processing {len(to_process)} products\n')
    
    found = not_found = errors = 0

    for i, row in enumerate(to_process):
        sku = row['SKU']
        print(f'[{i+1}/{len(to_process)}] {sku}... ', end='', flush=True)
        
        url = find_product_url(sku)
        
        if url:
            if args.dry_run:
                print(f'✓ DRY: {url.split("/ro/")[1][:60]}')
            else:
                ok = supabase_update(sku, url)
                if ok:
                    print(f'✓ {url.split("/ro/")[1][:60]}')
                    found += 1
                else:
                    print(f'✗ DB error')
                    errors += 1
        else:
            print('⚠ not found')
            not_found += 1
        
        # Polite delay between requests
        if i < len(to_process) - 1:
            time.sleep(1.5)

    print(f'\n── Summary ───────────────────────────')
    print(f'✓ Found:     {found}')
    print(f'⚠ Not found: {not_found}')
    print(f'✗ Errors:    {errors}')

if __name__ == '__main__':
    main()
